### PR TITLE
Query Builder

### PR DESCRIPTION
**Pull request by dantleech**
This PR adds a full query builder based on the PHPCR-ODM.

``` php
        $builder = $this->getDocumentManager()->createQueryBuilder();
        $builder->setLocale('en');
        $builder->from()->document('full', 'f');
        $builder->where()->eq()->field('f.status')->literal('closed');
```

``` php
        $builder->from('i')->joinInner()
            ->left()->document('full', 'f')->end()
            ->right()->document('issue', 'i')->end()
            ->condition()->equi('f.status', 'i.status');
        $results = $builder->getQuery()->execute();
```

It is based completely on the PHPCR-ODM query builder, the documentation for which can be found online.

### CODE REVIEW

File: lib/Query/ConverterSulu.php [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+use PHPCR\Query\QOM\QueryObjectModelConstantsInterface as QOMConstants;
+use PHPCR\SessionInterface;
+use Sulu\Component\DocumentManager\MetadataFactoryInterface;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Sulu\Component\DocumentManager\PropertyEncoder;
+
+/**
+ * Class which converts a Builder tree to a PHPCR Query.
+ */
+class ConverterSulu extends ConverterPhpcr
```
Comment (dantleech) [inline lib/Query/ConverterSulu.php]: Rename to BuilderConverter

File: lib/Query/ConverterSulu.php [hunk 2]
```diff
@@ -0,0 +1,260 @@
+<?php
+
```
Comment (dantleech) [inline lib/Query/ConverterSulu.php]: Run CS fixer.

File: lib/Query/ConverterSulu.php [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+     * @var Metadata[]
+     */
+    protected $documentMetadata = array();
+
+    /**
+     * @param PropertyEncoder
+     */
+    protected $encoder;
+
+    public function __construct(
```
Comment (dantleech) [inline lib/Query/ConverterSulu.php]: Doc

File: lib/Query/ConverterSulu.php [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+        );
+
+        $query = new Query($phpcrQuery, $this->eventDispatcher, $this->locale, array(), $builder->getPrimaryAlias());
+
+        if ($firstResult = $builder->getFirstResult()) {
+            $query->setFirstResult($firstResult);
+        }
+
+        if ($maxResults = $builder->getMaxResults()) {
+            $query->setMaxResults($maxResults);
```
Comment (dantleech) [inline lib/Query/ConverterSulu.php]: Add test for first / max.

File: lib/Query/ConverterSulu.php [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        }
+
+        return $query;
+    }
+
+    protected function applySourceConstraints(QueryBuilder $builder)
```
Comment (dantleech) [inline lib/Query/ConverterSulu.php]: Add `{@inheritDoc}` to all

File: lib/Query/ConverterSulu.php [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    {
+        // for each document source add phpcr:{class,classparents} restrictions
+        foreach ($this->sourceDocumentNodes as $sourceNode) {
+            $phpcrType = $this->documentMetadata[$sourceNode->getAlias()]->getPhpcrType();
+
+            // Jackalope-doctrine-dbal does not support selecting from mixins, and Sulu node types
+            // are currently mixins, so we need to explicitly add the mixin criteria.
+            $odmClassConstraints = $this->qomf->comparison(
```
Comment (dantleech) [inline lib/Query/ConverterSulu.php]: rename to `classConstraints`

File: lib/Query/ConverterSulu.php [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+                $this->qomf->propertyValue(
+                    $sourceNode->getAlias(),
+                    'jcr:mixinTypes'
```
Comment (dantleech) [inline lib/Query/ConverterSulu.php]: Should use the strategy

File: lib/Query/ConverterSulu.php [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+                QOMConstants::JCR_OPERATOR_EQUAL_TO,
+                $this->qomf->literal($phpcrType)
+            );
+
+            if ($this->constraint) {
+                $this->constraint = $this->qomf->andConstraint(
+                    $this->constraint,
+                    $odmClassConstraints
+                );
+            } else {
```
Comment (dantleech) [inline lib/Query/ConverterSulu.php]: Return early instead.

File: lib/Query/ConverterSulu.php [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+        foreach ($node->getChildren() as $property) {
+            list($alias, $phpcrName) = $this->getPhpcrProperty(
+                $property->getAlias(),
+                $property->getField()
+            );
+
+            $column = $this->qomf->column(
+                $alias,
+                $phpcrName,
+                // do we want to support custom column names in ODM?
```
Comment (dantleech) [inline lib/Query/ConverterSulu.php]: Can this be removed? Nothing here seems to be specific to Sulu.

File: lib/Query/ConverterSulu.php [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+
+        $this->documentMetadata[$alias] = $meta;
+
+        // NOTE: Currently we always select from [nt:unstructured] as Sulu node types
+        //       are mixins and jackalope-doctrine-dbal does not support selecting from mixins.
+        //
+        //       See note in getQuery
+        $alias = $this->qomf->selector(
+            $alias,
+            'nt:unstructured'
```
Comment (dantleech) [inline lib/Query/ConverterSulu.php]: Can this be moved to the strategy?

File: tests/Functional/Model/IssueDocument.php [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        return $this->name;
+    }
+    
+    public function setName($name)
+    {
+        $this->name = $name;
+    }
+}
+
```
Comment (dantleech) [inline tests/Functional/Model/IssueDocument.php]: EL

File: lib/DocumentStrategyInterface.php [hunk 12]
```diff
@@ -12,6 +12,9 @@
 namespace Sulu\Component\DocumentManager;
 
 use PHPCR\NodeInterface;
+use Sulu\Component\DocumentManager\Metadata;
+use PHPCR\Query\QOM\ConstraintInterface;
```
Comment (dantleech) [inline lib/DocumentStrategyInterface.php]: Remove

File: lib/MetadataFactoryInterface.php [hunk 13]
```diff
@@ -62,10 +62,10 @@ public function getMetadataForClass($class);
     /**
      * Return true if the document has metadata for the given fully qualified
      * class name.
-     *get_class($event->getDocument())get_class($event->getDocument())
+     *get_class($event->getDocument())get_class($event->getDocument()).
      t @param string
```
Comment (dantleech) [inline lib/MetadataFactoryInterface.php]: Remove

File: lib/Query/QueryBuilderConverter.php [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+        $this->applySourceConstraints($builder);
+
+        $phpcrQuery = $this->qomf->createQuery(
+            $this->from,
+            $this->constraint,
+            $this->orderings,
+            $this->columns
+        );
+
+        $query = new Query($phpcrQuery, $this->eventDispatcher, $this->locale, array(), $builder->getPrimaryAlias());
```
Comment (dantleech) [inline lib/Query/QueryBuilderConverter.php:126]: Option support can be added later (4th argument)

File: lib/Query/QueryBuilderConverter.php [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+
+        if ($firstResult = $builder->getFirstResult()) {
+            $query->setFirstResult($firstResult);
+        }
+
+        if ($maxResults = $builder->getMaxResults()) {
+            $query->setMaxResults($maxResults);
```
Comment (dantleech) [inline lib/Query/QueryBuilderConverter.php:133]: Still need to add tests for these.

File: lib/Query/QueryBuilderConverter.php [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+            $this->locale
+        );
+
+        return array($alias, $phpcrName);
+    }
+
+    /**
+     * Return either the metadata for the fqn of the document, or the alias
+     */
+    protected function getMetadata($documentFqn)
```
Comment (dantleech) [inline lib/Query/QueryBuilderConverter.php]: Complete doc.

File: lib/Strategy/MixinStrategy.php [hunk 17]
```diff
@@ -67,4 +69,26 @@ public function resolveMetadataForNode(NodeInterface $node)
...
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getPrimaryNodeType($classFqn)
+    {
+        return 'nt:unstructured';
+    }
+
+    public function createSourceConstraint(QueryObjectModelFactoryInterface $qomf, $sourceAlias, $classFqn)
```
Comment (dantleech) [inline lib/Strategy/MixinStrategy.php]: `{@inheritDoc}`

File: lib/Query/QueryBuilderConverter.php (hunk 15 above)
Comment (dantleech) [inline lib/Query/QueryBuilderConverter.php:133]: Added

**Comment by dantleech** [comment]
@danrot @wachterjohannes if you have time would be good if you review this, otherwise I will merge into `develop`.

File: lib/DocumentStrategyInterface.php [hunk 18]
```diff
@@ -48,4 +49,43 @@ public function createNodeForDocument($document, NodeInterface $parentNode, $nam
...
+    public function getPrimaryNodeType($classFqn);
+
+    /**
+     * Create a source constraint for a source document.
+     *
+     * For example:
+     *
+     * ````
+     * return $qomf->comparison(
+     *     $this->qomf->propertyValue(
```
Comment (dantleech) [inline lib/DocumentStrategyInterface.php]: s/this->//g

**Comment by dantleech** [comment]
Merged into develop. Feel free to review, I will make any fixes in develop.

File: lib/Query/QueryBuilderConverter.php [hunk 19]
```diff
@@ -0,0 +1,257 @@
...
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Sulu\Component\DocumentManager\DocumentStrategyInterface;
+
+/**
+ * Class which converts a Builder tree to a PHPCR Query.
+ */
+class QueryBuilderConverter extends ConverterPhpcr
+{
+    /**
+     * j
```
Comment (danrot) [inline lib/Query/QueryBuilderConverter.php:32]: Guess it shouldn't be here :-)

**Comment by danrot** [comment]
Are these `$qomf` variable names from somewhere else? Because it is an abbreviation which is not really common to me...

**Comment by dantleech** [comment]
I think `qomf` is an exception to the rule, it stands for `QueryObjectModelFactory`. The abbreviation is used both in Jackalope and PHPCR-ODM.

**Pull request merged**